### PR TITLE
Added iOS Audio Classifier and Basic Tests

### PR DESCRIPTION
**Pull request by priankakariatyml**
1. Added iOS Audio Classifier and Audio Tensor Utils.
2. Added basic classify objective c tests

### CODE REVIEW

**Comment by v7lin** [comment]
any news update?

File: tensorflow_lite_support/ios/test/task/audio/core/audio_record/utils/sources/AVAudioPCMBuffer+Utils.m [hunk 1]
```diff
@@ -80,4 +80,24 @@ + (nullable AVAudioPCMBuffer *)loadPCMBufferFromFileWithPath:(NSString *)path
...
+  AVAudioFormat *processingFormat =
+      [[AVAudioFormat alloc] initWithCommonFormat:AVAudioPCMFormatFloat32
+                                       sampleRate:audioFormat.sampleRate
+                                         channels:(AVAudioChannelCount)audioFormat.channelCount
+                                      interleaved:YES];
+
+  return [AVAudioPCMBuffer loadPCMBufferFromFileWithPath:path processingFormat:processingFormat];
+}
+
+- (nullable TFLFloatBuffer *)floatBuffer {
```
Comment (khanhlvg) [inline tensorflow_lite_support/ios/test/task/audio/core/audio_record/utils/sources/AVAudioPCMBuffer+Utils.m:95]: Is this used anywhere? If not, please remove.

File: tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.h [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+ *
+ * @param modelPath Path to the model file.
+ *
+ * @param error An optional error parameter populated if there is an error in initializing the audio
+ * classifier.
+ *
+ * @return A `TFLAudioClassifier` instance if initialization is successful or nil in case of
+ * failure.
+ */
+- (nullable instancetype)initWithModelPath:(NSString *)modelPath error:(NSError **)error;
```
Comment (khanhlvg) [inline tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.h:88]: Remove this to align with the ImageClassifier API.
https://github.com/tensorflow/tflite-support/blob/master/tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.h

File: tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.h [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+/**
+ * Initializes a new `TFLAudioClassifier` with the given `TFLAudioClassifierOptions`.
+ *
+ * @param options `TFLAudioClassifierOptions` initized with the absolute path to a model file stored
```
Comment (khanhlvg) [inline tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.h:93]: typo: initzed -> initialized

File: tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.h [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+ * locally on the device and other custom options for configuring the audio classifier.
+ *
+ * @param error An optional error parameter populated when there is an error in initializing the
+ * audio classifier.
+ *
+ * @return A `TFLAudioClassifier` instance if initialization is successful or nil in case of
+ * failure.
+ */
+- (nullable instancetype)initWithOptions:(TFLAudioClassifierOptions *)options
+                                   error:(NSError *_Nullable *)error;
```
Comment (khanhlvg) [inline tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.h:103]: (NSError **)error should be enough.

File: tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.h [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+ * Returns the required input buffer size in number elements or zero in case of failure to determine
+ * the input buffer size.
+ *
+ * @param error An optional error parameter populated if there is an error in determining the input
+ * buffer size.
+ *
+ * @return Required input buffer size in number elements or zero in case of failure to determine the
+ * input buffer size.
+ */
+- (NSInteger)requiredBufferSizeWithError:(NSError **)error;
```
Comment (khanhlvg) [inline tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.h:144]: As we already provided createInputAudioTensorWithError method to create the appropriate input audio tensor, we shouldn't expose requiredBufferSizeWithError and requiredTensorFormatWithError to avoid making the API unnecessarily complex.

File: tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.h [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+ * [documentation](https://www.tensorflow.org/lite/convert/metadata) for each task about the
+ * specific requirement.
+ *
+ * @param modelPath An absolute path to a TensorFlow Lite model file stored locally on the device.
+ *
+ * @return An instance of `TFLAudioClassifierOptions` initialized to the given model path.
+ */
+- (instancetype)initWithModelPath:(NSString *)modelPath;
+
+- (instancetype)init NS_UNAVAILABLE;
```
Comment (khanhlvg) [inline tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.h:58]: There's no need to mark these `init` and `new` method as NS_UNAVAILABLE. Let's align with the TFLImageClassifier API.

File: tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.m [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+  TFLAudioFormat *format = [[TFLAudioFormat alloc] initWithChannelCount:cFormat->channels
+                                                             sampleRate:cFormat->sample_rate];
+
+  TfLiteAudioFormatDelete(cFormat);
+
+  return format;
+}
+
+- (NSInteger)requiredBufferSizeWithError:(NSError **)error {
```
Comment (khanhlvg) [inline tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.m:175]: As this is a const and can be determined when the AudioClassifier is initialized, you should do it in one of the init method and store the value as an instance variable of this TFLAudioClassifier class. If this method failed for some reason, you can return an error in the initialization method.

By having requiredBufferSize as a const, you can simplify the `createInputAudioTensor` and `createAudioRecord` API by removing the `withError` parameter. The error path only happens when `requiredBufferSizeWithError:(NSError **)error` returns an error.

**Review by khanhlvg — CHANGES_REQUESTED** [review]

File: tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.m (hunk 7 above)
Comment (priankakariatyml) [reply tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.m]: Fixed

File: tensorflow_lite_support/ios/task/audio/core/audio_tensor/utils/sources/TFLAudioTensor+Utils.m [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ ==============================================================================*/
+#import "tensorflow_lite_support/ios/task/audio/core/audio_tensor/utils/sources/TFLAudioTensor+Utils.h"
+
+@implementation TFLAudioTensor (Utils)
+
+// Choosing to pass in a floatBuffer as opposed to using [self.ringBuffer floatBuffer]
+// to avoid creating an extra memcpy.
```
Comment (schmidt-sebastian) [inline tensorflow_lite_support/ios/task/audio/core/audio_tensor/utils/sources/TFLAudioTensor+Utils.m:20]: Can you rephrase this to make this a proper statement? To me, it seems like the beginning is missing.

File: tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.h (hunk 6 above)
Comment (schmidt-sebastian) [reply tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.h]: This doesn't seem to be resolved.

File: tensorflow_lite_support/ios/test/task/audio/audio_classifier/TFLAudioClassifierTests.m [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+  // audio format conversion differently as opposed to the task library C++ convenience method.
+  VerifyCategory(classificationResult.classifications[0].categories[2],
+                 485,          // expectedIndex
+                 0.003906,     // expectedScore
+                 @"Clicking",  // expectedLabel
+                 nil           // expectedDisplaName
+  );
+}
+
+- (void)testInferenceWithAudioRecordSucceeds {
```
[the remark this answers is not in the document]
Comment (priankakariatyml) [reply tensorflow_lite_support/ios/test/task/audio/audio_classifier/TFLAudioClassifierTests.m:174]: It is possible for some functionality. I have added a helper for one possible abstraction [here](https://github.com/khanhlvg/tflite-support/blob/ios-audio-classifier/tensorflow_lite_support/ios/test/task/audio/audio_classifier/TFLAudioClassifierTests.m#L185). Using helper methods for creating the classifier, audio tensor etc which are used by all methods will result in needing the created objects to be returned by the helpers. This may impact readability IMO since these are unit tests.

File: tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.h (hunk 6 above)
Comment (priankakariatyml) [reply tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.h]: It i resolved . Can you follow [this link](https://github.com/tensorflow/tflite-support/pull/835/files#diff-08fc61a126ede0c4b4e8476e20a6916e3641bfcc7426aeb01a3852ee1395c6d4R56) and check again. I think the link above this comment shows an outdated version.

File: tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.m [hunk 10]
```diff
@@ -52,43 +56,57 @@ - (instancetype)initWithModelPath:(NSString *)modelPath {
...
+
+  return _requiredCBufferSize > 0;
 }
 
-- (nullable instancetype)initWithCOptions:(TfLiteAudioClassifierOptions)cOptions
-                                    error:(NSError **)error {
-  TfLiteSupportError *cCreateClassifierError = NULL;
-  TfLiteAudioClassifier *cAudioClassifier =
-      TfLiteAudioClassifierFromOptions(&cOptions, &cCreateClassifierError);
+- (BOOL)populateRequiredTensorFormatWithError:(NSError **)error {
```
Comment (khanhlvg) [inline tensorflow_lite_support/ios/task/audio/sources/TFLAudioClassifier.m:76]: nit: Rename to `populateRequiredAudioFormatWithError` to better align with parameter name.

**Review by schmidt-sebastian — APPROVED** [review]

File: tensorflow_lite_support/ios/test/task/audio/audio_classifier/TFLAudioClassifierTests.m (hunk 9 above)
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply tensorflow_lite_support/ios/test/task/audio/audio_classifier/TFLAudioClassifierTests.m]: Looks great! Thank you
